### PR TITLE
Feature/jobs

### PR DESCRIPTION
**Pull request by RouHim**
## Rules for creators (remove these after creation)
- What's the story behind this PR?
- Every PR must have description of Impact / Risks / Unknowns

## Rules for reviewers
- Reject large PRs if it lacks detailed description / story behind it
- Large PRs may be done pairing with author
- "Approve" only if you approve of functional impact
- "Comment" only if you just did a code-style review
- Reject PRs with insufficient test cases
- **Meaningful rejections make us better**
- Optional: Check for cyclic maven dependencies

## Description
This PR:
- Solves #3 
- Adds a job queue
- Adds a scheduler that processes the enqueued jobs
- Store the job status in the database
- Adds test for the jobs feature

## Review
- [ ] functional review
- [ ] codestyle review

## Risks 
None

## Impact 
 - The crawling of the bistro page can atm just initiated by enqueuing a new crawler job.

## Unknowns
None

## Notes
Follow-Up ticket … created to clean … up.

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -1,4 +1,6 @@
 BISTRO_URL="https://bistro.cgm.ag/index.php"
 DATABASE_ADDRESS="http://localhost:8529"
 DATABASE_NAME="rate-by-bistro"
-COLLECTION_NAME="meals"
+MEAL_COLLECTION_NAME="meals"
```
Comment (ansgarS) [inline .env.example:4]: change to menus

File: .env.example [hunk 2]
```diff
@@ -1,4 +1,6 @@
 BISTRO_URL="https://bistro.cgm.ag/index.php"
 DATABASE_ADDRESS="http://localhost:8529"
 DATABASE_NAME="rate-by-bistro"
```
Comment (ansgarS) [inline .env.example:3]: change to bistro

File: .env.example [hunk 3]
```diff
@@ -1,4 +1,6 @@
...
-COLLECTION_NAME="meals"
+MEAL_COLLECTION_NAME="meals"
+JOB_COLLECTION_NAME="jobs"
+JOB_SCHEDULER_TICK_IN_SECONDS=10
```
Comment (ansgarS) [inline .env.example:6]: I think you have to add the db credentials, too

File: jobs/jobs.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+	if err != nil {
+		log.Fatal("Failed to init job scheduling", err)
+	}
+}
+
+// Gets called on every tick of the scheduler
+// It dequeues the head of the queue and start the parsing process.
+// Every job status change is persisted to the job collection.
+func processNextJob() {
```
Comment (ansgarS) [inline jobs/jobs.go:45]: This would not prevent multiple jobs running in parallel? Is this intended?

File: Main.go [hunk 5]
```diff
@@ -1,46 +1,26 @@
...
-type Config struct {
-	BistroUrl       string `env:"BISTRO_URL"`
-	DatabaseAddress string `env:"DATABASE_ADDRESS"`
-	DatabaseName    string `env:"DATABASE_NAME"`
-	CollectionName  string `env:"COLLECTION_NAME"`
+// the application cycle
+func main() {
+	jobs.EnqueueJob("2020-05-04")
+	jobs.EnqueueJob("2020-04-04")
+	jobs.EnqueueJob("2020-03-04")
```
Comment (ansgarS) [inline Main.go:13]: Why do you enqueue jobs here? Is it just for now and will be removed with an upcoming api?

File: config/config.go [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+	"log"
+)
+
+type Config struct {
+	BistroUrl                 string `env:"BISTRO_URL"`
+	DatabaseAddress           string `env:"DATABASE_ADDRESS"`
+	DatabaseName              string `env:"DATABASE_NAME"`
+	MealCollectionName        string `env:"MEAL_COLLECTION_NAME"`
+	JobCollectionName         string `env:"JOB_COLLECTION_NAME"`
+	JobSchedulerTickInSeconds uint64 `env:"JOB_SCHEDULER_TICK_IN_SECONDS"`
```
Comment (ansgarS) [inline config/config.go:15]: Why doesnt the JobScheduler simply pick the next job as soon as the last one is completed?

File: jobs/jobs.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+	"github.com/nu7hatch/gouuid"
+	"log"
+	"time"
+)
+
+// Represents a crawler job
+type Job struct {
+	Id          string   `json:"_key,omitempty"`
+	DateToParse string   `json:"dateToParse"`
+	Status      string   `json:"status"` // PENDING | RUNNING |  SUCCESS | FAILURE
```
Comment (ansgarS) [inline jobs/jobs.go:21]: I dont see a place where a job is persisted with status === failure

File: jobs/jobs.go [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+	Enqueued    string   `json:"enqueued"`
+	Started     string   `json:"started"`
+	Finished    string   `json:"finished"`
```
[the remark this answers is not in the document]
Comment (ansgarS) [reply jobs/jobs.go:24]: These are dates, correct?

File: jobs/.env.example [hunk 9]
```diff
@@ -0,0 +1,4 @@
+DATABASE_ADDRESS="http://localhost:8529"
```
[the remark this answers is not in the document]
Comment (RouHim) [reply jobs/.env.example:1]: No, this issue is resolved in the next feature (rest)

File: jobs/jobs.go [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+	"github.com/go-co-op/gocron"
+	"github.com/nu7hatch/gouuid"
+	"log"
+	"time"
+)
+
+// Represents a crawler job
+type Job struct {
+	Id          string   `json:"_key,omitempty"`
+	DateToParse string   `json:"dateToParse"`
```
[the remark this answers is not in the document]
Comment (RouHim) [reply jobs/jobs.go:20]: this field represents the day that the crawler have to parse.

File: jobs/jobs.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (RouHim) [reply jobs/jobs.go:24]: Jup i will rename them to be clearer

File: .env.example (hunk 1 above)
Comment (RouHim) [reply .env.example:4]: also resolved in the next  feature.

File: .env.example (hunk 2 above)
Comment (RouHim) [reply .env.example:3]: is fixed in the next feature (rest)

File: .env.example (hunk 3 above)
Comment (RouHim) [reply .env.example:6]: also fixed in the next feature (rest)
this is a pretty old PR.

File: Main.go (hunk 5 above)
Comment (RouHim) [reply Main.go:13]: Yep you got it.

File: jobs/jobs.go (hunk 7 above)
Comment (RouHim) [reply jobs/jobs.go:21]: Thats an issue, fixing....

Comment (RouHim) [reply jobs/jobs.go:21]: I will create an bug issue ticket (#15) for that an fix it in the based on the next feature/rest to ensure a correct error reporting via the rest api.

File: config/config.go (hunk 6 above)
Comment (RouHim) [reply config/config.go:15]: Can be an option, but there have to be a way to start a job when the queue is empty. If the queue is empty and a job is enqueued there is no previous one to start the new one.

File: jobs/jobs.go (hunk 4 above)
Comment (RouHim) [reply jobs/jobs.go:45]: Good point, but I don't think it's a problem to get rid of the queue sequentially and then process it in parallel.

File: jobs/.env.example [hunk 11]
```diff
@@ -0,0 +1,4 @@
...
+DATABASE_NAME="rate-by-bistro_test"
```
Comment (ansgarS) [inline jobs/.env.example:2]:
```suggestion
DATABASE_NAME="rate-my-bistro_test"
```

**Pull request merged**